File: core/src/main/java/com/starshootercity/geysermc/GeyserSwapper.java
package com.starshootercity.geysermc;

import com.starshootercity.*;
import com.starshootercity.commands.OriginCommand;
import com.starshootercity.events.PlayerSwapOriginEvent;
import net.kyori.adventure.text.Component;
import net.milkbowl.vault.economy.Economy;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.geysermc.cumulus.form.Form;
import org.geysermc.cumulus.form.ModalForm;
import org.geysermc.cumulus.form.SimpleForm;
import org.geysermc.cumulus.util.FormImage;
import org.geysermc.floodgate.api.FloodgateApi;
import org.geysermc.geyser.api.GeyserApi;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.UUID;

import static com.starshootercity.OriginSwapper.orbCooldown;

public class GeyserSwapper {
    public static boolean checkBedrockSwap(Player player, PlayerSwapOriginEvent.SwapReason reason, boolean cost, boolean displayOnly) {
        try {
            if (!ShortcutUtils.isBedrockPlayer(player.getUniqueId())) {
                return true;
            } else {
                openOriginSwapper(player, reason, displayOnly, cost);
                return false;
            }
        } catch (NoClassDefFoundError e) {
            return true;
        }
    }

    public static void openOriginSwapper(Player player, PlayerSwapOriginEvent.SwapReason reason, boolean displayOnly, boolean cost) {
        List<Origin> origins = new ArrayList<>(AddonLoader.origins);
        if (!displayOnly) origins.removeIf(origin -> origin.isUnchoosable(player) || origin.hasPermission() && !player.hasPermission(origin.getPermission()));
        else {
            openOriginInfo(player, OriginSwapper.getOrigin(player), PlayerSwapOriginEvent.SwapReason.COMMAND, true, false);
            return;
        }
        origins.sort((o1, o2) -> {
            if (o1.getImpact() == o2.getImpact()) {
                if (o1.getPosition() == o2.getPosition()) return 0;
                return o1.getPosition() > o2.getPosition() ? 1 : -1;
            }
            return o1.getImpact() > o2.getImpact() ? 1 : -1;
        });

        SimpleForm.Builder form = SimpleForm.builder().title("Choose your Origin");

        for (Origin origin : origins) {
            if (origin.getIcon().getType() == Material.PLAYER_HEAD) {
                form.button(origin.getName(), FormImage.Type.URL, "https://mc-heads.net/avatar/MHF_Steve");
            } else {
                form.button(origin.getName(), FormImage.Type.URL, origin.getResourceURL());
            }

        }

        boolean enableRandom = OriginsReborn.getInstance().getConfig().getBoolean("origin-selection.random-option.enabled");
        if (enableRandom) form.button("Random", FormImage.Type.URL, "https://static.wikia.nocookie.net/origins-smp/images/1/13/Origin_Orb.png/revision/latest?cb=20210411202749");

        sendForm(player.getUniqueId(), form
                .closedOrInvalidResultHandler(() -> {
                    if (OriginSwapper.getOrigin(player) == null) {
                        openOriginSwapper(player, reason, false, cost);
                    }
                })
                .validResultHandler(response -> openOriginInfo(player, AddonLoader.originNameMap.get(response.clickedButton().text().toLowerCase()), reason, false, cost)).build());

    }

    private static void sendForm(UUID uuid, Form form) {
        try {
            FloodgateApi.getInstance().sendForm(uuid, form);
        } catch (NoClassDefFoundError e) {
            GeyserApi.api().sendForm(uuid, form);
        }
    }

    private static final Random random = new Random();

    public static void setOrigin(Player player, Origin origin, PlayerSwapOriginEvent.SwapReason reason, boolean cost) {
        if (OriginsReborn.getInstance().isVaultEnabled() && cost) {
            int amount = OriginsReborn.getInstance().getConfig().getInt("swap-command.vault.cost", 1000);
            if (origin.getCost() != null) amount = origin.getCost();
            Economy economy = OriginsReborn.getInstance().getEconomy();
            if (economy.has(player, amount)) {
                economy.withdrawPlayer(player, amount);
            } else {
                String symbol = OriginsReborn.getInstance().getConfig().getString("swap-command.vault.currency-symbol", "$");
                player.sendMessage(Component.text("You need %s%s to swap your origin!".formatted(symbol, amount)));
                return;
            }
        }


        ItemMeta meta = player.getInventory().getItemInMainHand().getItemMeta();

        if (reason == PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN) {
            EquipmentSlot hand = null;
            if (meta != null) {
                if (meta.getPersistentDataContainer().has(OrbOfOrigin.orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) {
                    hand = EquipmentSlot.HAND;
                }
            }
            if (hand == null) {
                ItemMeta offhandMeta = player.getInventory().getItemInOffHand().getItemMeta();
                if (offhandMeta != null) {
                    if (offhandMeta.getPersistentDataContainer().has(OrbOfOrigin.orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) {
                        hand = EquipmentSlot.OFF_HAND;
                    }
                }
            }
            if (hand == null) return;
            orbCooldown.put(player, System.currentTimeMillis());
            if (hand == EquipmentSlot.HAND) player.swingMainHand();
            else player.swingOffHand();
            if (OriginsReborn.getInstance().getConfig().getBoolean("orb-of-origin.consume")) {
                ItemStack i = player.getInventory().getItem(hand);
                if (i != null) i.setAmount(i.getAmount() - 1);
            }
        }
        boolean resetPlayer = OriginSwapper.shouldResetPlayer(reason);
        if (origin == null) {
            List<Origin> origins = new ArrayList<>(AddonLoader.origins);
            origins.removeIf(origin1 -> origin1.isUnchoosable(player));
            List<String> excludedOrigins = OriginsReborn.getInstance().getConfig().getStringList("origin-selection.random-option.exclude");
            origins.removeIf(possibleOrigin -> excludedOrigins.contains(possibleOrigin.getName()));
            origin = origins.get(random.nextInt(origins.size()));
        }
        OriginsReborn.getCooldowns().setCooldown(player, OriginCommand.key);
        OriginSwapper.setOrigin(player, origin, reason, resetPlayer);
    }

    public static void openOriginInfo(Player player, Origin origin, PlayerSwapOriginEvent.SwapReason reason, boolean displayOnly, boolean cost) {
        ModalForm.Builder form = ModalForm.builder();
        StringBuilder info = new StringBuilder();
        if (origin != null) {
            form.title(origin.getName());
            for (OriginSwapper.LineData.LineComponent line : new OriginSwapper.LineData(origin).getRawLines()) {
                if (line.isEmpty()) {
                    info.append("\n\n");
                } else {
                    info.append(line.getRawText());
                }
                if (line.getType() == OriginSwapper.LineData.LineComponent.LineType.TITLE) info.append("\n");
            }
        } else {
            form.title("Random Origin");
            List<Origin> origins = new ArrayList<>(AddonLoader.origins);
            origins.removeIf(origin1 -> origin1.isUnchoosable(player));
            List<String> excludedOrigins = OriginsReborn.getInstance().getConfig().getStringList("origin-selection.random-option.exclude");
            info.append("You'll be assigned one of the following:\n\n");
            for (Origin possibleOrigin : origins) {
                if (!excludedOrigins.contains(possibleOrigin.getName())) {
                    info.append(possibleOrigin.getName()).append("\n");
                }
            }
        }
        if (cost) {
            String symbol = OriginsReborn.getInstance().getConfig().getString("swap-command.vault.currency-symbol", "$");
            int amount = OriginsReborn.getInstance().getConfig().getInt("swap-command.vault.cost", 1000);
            if (origin != null) {
                if (origin.getCost() != null) amount = origin.getCost();
            }
            info.append("\n\n\n§eThis will cost you %s%s!".formatted(symbol, amount));
        }
        form.content(info.toString());
        form.button1("Cancel");
        form.button2("Confirm");
        sendForm(player.getUniqueId(), form
                .closedOrInvalidResultHandler(() -> {
                    if (!displayOnly) {
                        openOriginInfo(player, origin, reason, false, cost);
                    }
                })
                .validResultHandler(response -> {
                    if (displayOnly) return;
                    if (response.clickedButtonId() == 0) {
                        openOriginSwapper(player, reason, false, cost);
                    } else {
                        setOrigin(player, origin, reason, cost);
                    }
                }).build());
    }
}

File: core/src/main/java/com/starshootercity/geysermc/GeyserSwapper.kt
package com.starshootercity.geysermc

import com.starshootercity.AddonLoader.getOrigin
import com.starshootercity.AddonLoader.getOrigins
import com.starshootercity.OrbOfOrigin
import com.starshootercity.Origin
import com.starshootercity.OriginSwapper
import com.starshootercity.OriginSwapper.Companion.getOrigin
import com.starshootercity.OriginSwapper.Companion.shouldResetPlayer
import com.starshootercity.OriginSwapper.LineData
import com.starshootercity.OriginsReborn
import com.starshootercity.OriginsReborn.Companion.getCooldowns
import com.starshootercity.OriginsReborn.Companion.instance
import com.starshootercity.ShortcutUtils.isBedrockPlayer
import com.starshootercity.commands.OriginCommand
import com.starshootercity.events.PlayerSwapOriginEvent.SwapReason
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.EquipmentSlot
import org.endera.enderalib.utils.async.ioDispatcher
import org.geysermc.cumulus.form.Form
import org.geysermc.cumulus.form.ModalForm
import org.geysermc.cumulus.form.SimpleForm
import org.geysermc.cumulus.util.FormImage
import org.geysermc.floodgate.api.FloodgateApi
import org.geysermc.geyser.api.GeyserApi
import java.util.*

object GeyserSwapper {

    suspend fun checkBedrockSwap(
        player: Player,
        reason: SwapReason,
        cost: Boolean,
        displayOnly: Boolean,
        layer: String
    ): Boolean {
        try {
            if (!isBedrockPlayer(player.uniqueId)) {
                return true
            } else {
                openOriginSwapper(player, reason, displayOnly, cost, layer)
                return false
            }
        } catch (_: NoClassDefFoundError) {
            return true
        }
    }

    suspend fun openOriginSwapper(
        player: Player,
        reason: SwapReason,
        displayOnly: Boolean,
        cost: Boolean,
        layer: String
    ) {
        if (displayOnly) {
            getOrigin(player, layer)?.let { origin ->
                openOriginInfo(
                    player = player,
                    origin = origin,
                    reason = SwapReason.COMMAND,
                    displayOnly = true,
                    cost = false,
                    layer = layer
                )
            }
            return
        }

        val origins = getOrigins(layer)
            .filterNot { origin ->
                origin.isUnchoosable(player) ||
                        (origin.hasPermission() && (origin.permission == null || !player.hasPermission(origin.permission)))
            }
            .sortedWith(compareBy({ it.impact }, { it.position }))

        val form = SimpleForm.builder().apply {
            title("Choose your Origin")
            origins.forEach { origin ->
                val imageUrl = if (origin.icon.type == Material.PLAYER_HEAD)
                    "https://mc-heads.net/avatar/MHF_Steve"
                else
                    origin.getResourceURL()
                button(origin.getName(), FormImage.Type.URL, imageUrl)
            }
            if (instance.getConfig().getBoolean("origin-selection.random-option.enabled")) {
                button(
                    "Random",
                    FormImage.Type.URL,
                    "https://static.wikia.nocookie.net/origins-smp/images/1/13/Origin_Orb.png/revision/latest?cb=20210411202749"
                )
            }
        }

        sendForm(
            player.uniqueId,
            form
                .closedOrInvalidResultHandler(Runnable {
                    CoroutineScope(ioDispatcher).launch {
                        // If no origin is chosen, reopen the swapper.
                        if (getOrigin(player, layer) == null) {
                            openOriginSwapper(player, reason, displayOnly = false, cost = cost, layer = layer)
                        }
                    }
                })
                .validResultHandler { response ->
                    val clickedOrigin = getOrigin(response.clickedButton().text().lowercase(Locale.getDefault()))
                    CoroutineScope(ioDispatcher).launch {
                        openOriginInfo(
                            player = player,
                            origin = clickedOrigin,
                            reason = reason,
                            displayOnly = false,
                            cost = cost,
                            layer = layer
                        )
                    }
                }
                .build()
        )
    }

    private fun sendForm(uuid: UUID, form: Form) {
        try {
            FloodgateApi.getInstance().sendForm(uuid, form)
        } catch (_: NoClassDefFoundError) {
            GeyserApi.api().sendForm(uuid, form)
        }
    }

    private val random = Random()

    suspend fun setOrigin(
        player: Player,
        origin: Origin?,
        reason: SwapReason,
        cost: Boolean,
        layer: String
    ) {
        var selectedOrigin = origin

        if (instance.isVaultEnabled && cost) {
            val defaultCost = OriginsReborn.mainConfig.swapCommand.vault.defaultCost
            val amount = selectedOrigin?.cost ?: defaultCost
            val amountDouble = amount.toDouble()
            val economy = checkNotNull(instance.economy)
            if (economy.has(player, amountDouble)) {
                economy.withdrawPlayer(player, amountDouble)
            } else {
                val symbol = OriginsReborn.mainConfig.swapCommand.vault.currencySymbol
                player.sendMessage(Component.text("You need $symbol$amount to swap your origin!"))
                return
            }
        }

        if (reason == SwapReason.ORB_OF_ORIGIN) {
            val hand = when {
                player.inventory.itemInMainHand.itemMeta?.persistentDataContainer
                    ?.has(OrbOfOrigin.orbKey, OriginSwapper.BooleanPDT.BOOLEAN) == true -> EquipmentSlot.HAND
                player.inventory.itemInOffHand.itemMeta?.persistentDataContainer
                    ?.has(OrbOfOrigin.orbKey, OriginSwapper.BooleanPDT.BOOLEAN) == true -> EquipmentSlot.OFF_HAND
                else -> null
            } ?: return

            OriginSwapper.orbCooldown[player] = System.currentTimeMillis()
            withContext(OriginsReborn.bukkitDispatcher) {
                when (hand) {
                    EquipmentSlot.HAND -> player.swingMainHand()
                    EquipmentSlot.OFF_HAND -> player.swingOffHand()
                    else -> {}
                }
                if (OriginsReborn.mainConfig.orbOfOrigin.consume) {
                    player.inventory.getItem(hand)?.let { item ->
                        item.amount = item.amount - 1
                    }
                }
            }
        }

        val resetPlayer = shouldResetPlayer(reason)
        if (selectedOrigin == null) {
            val excludedOrigins = OriginsReborn.mainConfig.originSelection.randomOption.exclude
            val origins = getOrigins(layer)
                .filter { !it.isUnchoosable(player) && excludedOrigins.contains(it.getName())}
            selectedOrigin = origins[random.nextInt(origins.size)]
        }

        getCooldowns().setCooldown(player, OriginCommand.key)
        OriginSwapper.setOrigin(player, selectedOrigin, reason, resetPlayer, layer)
    }

    suspend fun openOriginInfo(
        player: Player,
        origin: Origin?,
        reason: SwapReason,
        displayOnly: Boolean,
        cost: Boolean,
        layer: String
    ) {
        val form = ModalForm.builder()
        val info = StringBuilder()

        if (origin != null) {
            form.title(origin.getName())
            LineData(origin).rawLines.forEach { line ->
                checkNotNull(line)
                info.append(if (line.isEmpty) "\n\n" else line.rawText)
                if (line.type == LineData.LineComponent.LineType.TITLE) info.append("\n")
            }
        } else {
            form.title("Random Origin")
            val excludedOrigins = OriginsReborn.mainConfig.originSelection.randomOption.exclude
            val origins = getOrigins(layer).filter { !it.isUnchoosable(player) }
            info.append("You'll be assigned one of the following:\n\n")
            origins.filter { it.getName() !in excludedOrigins }
                .forEach { possibleOrigin -> info.append(possibleOrigin.getName()).append("\n") }
        }

        if (cost) {
            val symbol = OriginsReborn.mainConfig.swapCommand.vault.currencySymbol
            var amount = OriginsReborn.mainConfig.swapCommand.vault.defaultCost
            origin?.cost?.let { amount = it }
            info.append("\n\n\n§eThis will cost you $symbol$amount!")
        }

        form.content(info.toString())
            .button1("Cancel")
            .button2("Confirm")

        sendForm(
            player.uniqueId, form
                .closedOrInvalidResultHandler(Runnable {
                    if (!displayOnly) {
                        CoroutineScope(ioDispatcher).launch {
                            openOriginInfo(player, origin, reason, false, cost, layer)
                        }
                    }
                })
                .validResultHandler { response ->
                    if (displayOnly) return@validResultHandler
                    CoroutineScope(ioDispatcher).launch {
                        if (response?.clickedButtonId() == 0) {
                            openOriginSwapper(player, reason, false, cost, layer)
                        } else {
                            setOrigin(player, origin, reason, cost, layer)
                        }
                    }
                }
                .build()
        )
    }

}
